### PR TITLE
feat: Dependent Fields: Refactoring

### PR DESCRIPTION
**Pull request by yksolanki9**
Clickup - https://app.clickup.com/t/85zrrve55

### CODE REVIEW

File: src/app/fyle/add-edit-expense/add-edit-expense.page.html [hunk 1]
```diff
@@ -402,12 +402,20 @@
...
           <ng-container *ngIf="txnFields$|async as txnFields">
             <ng-container *ngIf="dependentFields$|async as dependentFields">
-              <app-dependent-fields
-                #dependentFields
-                [expenseForm]="fg"
-                [txnFields]="txnFields"
-                [dependentCustomFields]="dependentFields"
-              ></app-dependent-fields>
+              <ng-container *ngIf="etxn$|async as etxn">
+                <ng-container *ngIf="selectedProject$|async as selectedProject">
```
Comment (yksolanki9) [inline src/app/fyle/add-edit-expense/add-edit-expense.page.html:406]: `selectedProject$` is a BehaviourSubject - added more details below

File: src/app/fyle/add-edit-expense/add-edit-expense.page.ts [hunk 2]
```diff
@@ -348,6 +349,8 @@ export class AddEditExpensePage implements OnInit {
 
   dependentFields$: Observable<ExpenseField[]>;
 
+  selectedProject$: BehaviorSubject<ExtendedProject>;
```
Comment (rvab) [inline src/app/fyle/add-edit-expense/add-edit-expense.page.ts:352]: I'm not sure if this is a good approach. Right now we can go ahead with it.

File: src/app/shared/components/dependent-fields/dependent-fields.component.ts [hunk 3]
```diff
@@ -37,37 +29,25 @@ export class DependentFieldsComponent implements OnInit, OnDestroy {
...
+        value: this.parentFieldValue,
       });
+    }
   }
 
   ngOnDestroy() {
-    this.onPageExit$.next(null);
-    this.onPageExit$.complete();
+    this.onPageExit$?.next(null);
+    this.onPageExit$?.complete();
```
Comment (rvab) [inline src/app/shared/components/dependent-fields/dependent-fields.component.ts:50]: When will `this.onPageExit$` be undefined?

**Review by rvab — APPROVED** [review]
Minor comments. Apart from selectedProject$, rest lgtm

File: src/app/shared/components/dependent-fields/dependent-fields.component.ts [hunk 4]
```diff
@@ -37,37 +29,25 @@ export class DependentFieldsComponent implements OnInit, OnDestroy {
...
-      )
-      .subscribe((dependentFieldDetails) => {
-        if (dependentFieldDetails?.dependentField) {
-          this.addDependentField(dependentFieldDetails.dependentField, dependentFieldDetails.parentFieldValue);
-        }
+  }
+
+  ngOnChanges(changes: SimpleChanges) {
+    if (changes?.parentFieldValue) {
+      this.onPageExit$ = new Subject();
```
Comment (Julias0) [inline src/app/shared/components/dependent-fields/dependent-fields.component.ts:38]: why is this initialized inside ngOnChanges? Why not do it inside ngOnInit?

**Review by Julias0 — CHANGES_REQUESTED** [review]
1 comment - rest lgtm

File: src/app/shared/components/dependent-fields/dependent-fields.component.ts (hunk 4 above)
Comment (yksolanki9) [reply src/app/shared/components/dependent-fields/dependent-fields.component.ts:38]: ngOnChanges is called before ngOnInit. And the method inside ngOnChanges uses this subject, which is undefined when ngOnChanges is executed for the first time, so have initialized it in ngOnChanges as well

File: src/app/shared/components/dependent-fields/dependent-fields.component.ts (hunk 3 above)
Comment (yksolanki9) [reply src/app/shared/components/dependent-fields/dependent-fields.component.ts:50]: Not sure how it happened, but saw an error where this was undefined, so added optional chaining

File: src/app/shared/components/dependent-fields/dependent-fields.component.ts (hunk 4 above)
Comment (yksolanki9) [reply src/app/shared/components/dependent-fields/dependent-fields.component.ts:36]: Have removed the initialization from `ngOnInit()` and triggering initializing it in ngOnChanges only once using firstChange

**Review by Julias0 — APPROVED** [review]

**Pull request merged**